edge-modules/azure-monitor/src/Program.cs

```diff
@@ -89,5 +89,21 @@ static async Task<int> MainAsync()
             LoggerUtil.Writer.LogInformation("MetricsCollector Main() finished.");
             return 0;
         }
+
+        private static Func<Tuple<Metric, string>, Tuple<Metric, string>> ProvideIdentifyingTags()
```

Comment:
@alaendle Thank you for taking the time to put in this PR. 

After looking through your changes, I think that it will be simpler and with no modification to the method signatures if we moved this logic to the Prometheus ParseMetrics function at this line: https://github.com/Azure/iotedge/blob/e8d8240253fd445f54dfd2c422d2a0a47761e8a9/edge-modules/azure-monitor/src/PrometheusMetricsParser.cs#L144

Here, you could do something like this to augment the missing values in the metric.
~~~
if (Settings.Current.AddIdentifyingTags)
{
       if (!tags.ContainsKey("edge_device")) tags["edge_device"] = Environment.GetEnvironmentVariable("IOTEDGE_DEVICEID");
       if (!tags.ContainsKey("iothub")) tags["iothub"] = Environment.GetEnvironmentVariable("IOTEDGE_IOTHUBHOSTNAME");
       if (!tags.ContainsKey("module_name")) tags["module_name"] = new Uri(endpoint).Host;
}
~~~

Also, regarding your question about label naming, the names you have chosen here are the right ones. I cross referenced these names with our tests and other areas of the code and validated that they are consistent.

Reply:
@scb01 Thanks for looking at the changes. I've adjusted the PR accordingly - only with the small change to check `endpoint` for `null`.

Reply:
@alaendle Good call on adding the check for `null`. Thinking along the same lines, perhaps we should put this logic in a try catch to protect against any (unlikely) exceptions that are thrown by `Environment.GetEnvironmentVariable` and `new Uri` methods.

if an exception does occur, we could just simply log a warning and carry on, so as to minimize any disruption to the existing logic.

Reply:
@scb01 I've added a very simple/generic catch clause. Personally I'm not a fan of these generic error handling routines - e.g. we would now also "handle" an OutOfMemoryException - that's why nothing like that was present in the first place. However I can fully understand your intention to minimize possible impact of this code addition.